### PR TITLE
Add unwrap_var function to remove unsafe wrappers

### LINKED ISSUE #70357: AnsibleUnsafeText not casted to basestring when calling from_yaml
##### SUMMARY

Using a Jinja2 expression using `lookup()` and `from_yaml` produces an error, if the C-Extension of PyYAML is used.
The reason seems to be that Ansible's `from_yaml`-filter calls the YAML-parser with an instance of `AnsibleUnsafeText`
The C-extension of PyYAML explicitly denies subclasses of string/basestring: https://github.com/yaml/pyyaml/blob/23c952fe08b2e7ea0f8d7673f45b17547e331f4b/ext/_yaml.pyx#L299

See also https://bugs.launchpad.net/ubuntu/+source/ansible/+bug/1880226

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

core

##### ANSIBLE VERSION

```paste below
ansible 2.9.9
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Apr  2 2020, 13:16:51) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]
```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT

CentOS 7
In my case Ansible only uses the C implentation of PyYAML, if `salt` has been installed from https://repo.saltstack.com/yum/redhat/7/x86_64/3000/
I do not use the PyYAML package from the saltstack-repository and even if I do it does not work.
As soon as I uninstall the `salt`-package, Ansible uses the native-python YAML-parser of PyYAML and the filter works.

##### STEPS TO REPRODUCE

It is important that Ansible uses PyYAML with the `libyaml` C-library.
If it uses the YAML-parser written in plain python the error does not occur.
For me this was -strangely- only the case if `salt` was installed (see OS / ENVIRONMENT)

```yaml
---
- hosts: localhost
  tasks:
  - debug:
      msg: "{{ lookup('file', 'testfile.yaml') | from_yaml}}"
...
```

testfile.yaml:
```yaml
---
Hello: World
...
```

##### EXPECTED RESULTS

The file content read by `lookup()` should be accepted by the `from_yaml`-filter.

##### ACTUAL RESULTS

```paste below
TASK [debug] **********************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"msg": "Unexpected templating type error occurred on ({{ lookup('file', 'testfile.yaml') | from_yaml}}): a string or stream input is required"}
```

### PR DESCRIPTION
**Pull request by sivel**
##### SUMMARY
Fixes #70357
Add `unwrap_var` function to remove unsafe wrappers

Use of this new function should be strictly monitored, to ensure proper use.

TODO:

- [x] changelog
- [x] CI step to fail PRs that make use of `unwrap_var` unless some explicit acceptance is performed.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
lib/ansible/utils/unsafe_proxy.py

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by sivel** [comment]
cc @jborean93 @abadger

File: lib/ansible/utils/unsafe_proxy.py [hunk 1]
```diff
@@ -115,26 +115,63 @@ def _wrap_set(v):
...
+    Python API where the receiving python code does not understand our subclasses, potentially due to
+    performing type checks like ``if type(value) == 'str':``
+
+    This is not required to convert values to JSON, as the default JSON encoder will implicitly ignore the
+    subclass.
+    """
+    if isinstance(v, AnsibleUnsafeBytes):
+        return binary_type(v)
+    elif isinstance(v, AnsibleUnsafeText):
+        return text_type(v)
```
Comment (abadger) [inline lib/ansible/utils/unsafe_proxy.py:168]: I think, to future proof, we should add an `elif isinstance(v, string_types): return v` here.

The possible future it protects against is if we want to handle immutable sequences.  Then we have to make sure we've gotten rid of both wrapped and unwrapped string types before then

File: test/units/utils/test_unsafe_proxy.py [hunk 2]
```diff
@@ -85,3 +85,37 @@ def test_AnsibleUnsafeText():
...
+                AnsibleUnsafeText(u'text'),
+                AnsibleUnsafeBytes(b'bytes'),
+            )
+        ),
+        'tuple': (
+            AnsibleUnsafeText(u'text'),
+            AnsibleUnsafeBytes(b'bytes'),
+        )
+    }
+
```
Comment (abadger) [inline test/units/utils/test_unsafe_proxy.py:139]: Maybe mix some unwrapped text and byte strings in here as well to make sure we don't get caught out by that falling through to an immutable sequence check in some future iteration)

Also, i like to always include some non-ascii data when testing that arbitrary strings are doing the right thing.

File: lib/ansible/utils/unsafe_proxy.py (hunk 1 above)
Comment (sivel) [reply lib/ansible/utils/unsafe_proxy.py:168]: I can live with that, but do you think we need `elif isinstance(v, (binary_type, text_type)):` instead?

Comment (abadger) [reply lib/ansible/utils/unsafe_proxy.py:168]: Good idea!  I like your new code.

**Comment by abadger** [comment]
Looking good.  +1 from me.

File: lib/ansible/utils/unsafe_proxy.py [hunk 3]
```diff
@@ -114,26 +114,67 @@ def _wrap_set(v):
...
+        return _wrap_dict(v)
     elif is_sequence(v):
-        v = _wrap_sequence(v)
-    elif isinstance(v, binary_type):
-        v = AnsibleUnsafeBytes(v)
-    elif isinstance(v, text_type):
-        v = AnsibleUnsafeText(v)
+        return _wrap_sequence(v)
 
     return v
```
Comment (samdoran) [inline lib/ansible/utils/unsafe_proxy.py:141]: Do we still need this return statement here? What types would fall through to this return?

**Review by samdoran — APPROVED** [review]

File: lib/ansible/utils/unsafe_proxy.py (hunk 3 above)
Comment (sivel) [reply lib/ansible/utils/unsafe_proxy.py:141]: `bool`, `int`, `RandomType`...

**Comment by sivel** [comment]
fwiw, I included the `yaml` utils to prefer the C extension in here, largely because that change needs this one.

**Comment by samdoran** [comment]
> fwiw, I included the yaml utils to prefer the C extension in here

I was hoping we could do something like this to make it easy to get the C extension without having to reinvent the wheel everywhere. I like it.

**Review by bcoca — APPROVED** [review]

**Review by Shrews — APPROVED** [review]

**Comment by nitzmahone** [comment]
If we're going to do the unsafe revisit in 2.11, do we want to merge this now?

**Comment by sivel** [comment]
> If we're going to do the unsafe revisit in 2.11

Are we? And I have no idea what that means in any real context.

File: test/lib/ansible_test/_data/sanity/code-smell/unwrap-unsafe.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        found = 0
+        with open(path, 'r') as f:
+            for line in f.readlines():
+                matches = UNWRAP_RE.findall(line)
+                if matches:
+                    found += 1
+
+        if found > WHITELIST.get(path, 0):
+            print(
+                '%s: use of unwrap_var is strictly monitored for incorrect use. Please seek guidance from a core committer' % (path,)
```
Comment (mattclay) [inline test/lib/ansible_test/_data/sanity/code-smell/unwrap-unsafe.py:29]: This test will run on both ansible-base and collections, but this message (and the list to bypass it) aren't really an option for collection authors. What's the intended resolution in those cases?

Comment (sivel) [reply test/lib/ansible_test/_data/sanity/code-smell/unwrap-unsafe.py:29]: This is a complicated question, one in which I would hope that collection authors aren't using. My plan was just that collection authors would have to add an ignore.

Comment (mattclay) [reply test/lib/ansible_test/_data/sanity/code-smell/unwrap-unsafe.py:29]: In that case I think the phrasing should be different for ansible-base vs collections so relevant instructions can be given in each context.

**Review by samdoran — APPROVED** [review]

**Comment by sivel** [comment]
Closing since we are re-implementing unsafe.

**Pull request closed without merging**